File: Arithmetic_&_Basic_Reasoning/Q3. Palindrome Number/Q3_PalindromeNumber.java
class Solution {
    public boolean isPalindrome(int x) {
        // Negative numbers are not palindromes
        // Numbers ending with 0 (except 0 itself) can't be palindromes
        if (x < 0 || (x % 10 == 0 && x != 0)) {
            return false;  // directly return false
        }
        
        int reversedHalf = 0;  // to store the reversed half of the number
        
        // Keep reversing the last digits until we reach the middle
        while (x > reversedHalf) {
            int digit = x % 10;                     // get the last digit
            reversedHalf = reversedHalf * 10 + digit; // add digit to reversed half
            x /= 10;                                // remove the last digit from x
        }
        
        // If number has even digits: both halves will be equal
        // If number has odd digits: ignore the middle digit by dividing reversedHalf by 10
        return x == reversedHalf || x == reversedHalf / 10;
    }
}

File: Arithmetic_&_Basic_Reasoning/Q3. Palindrome Number/Q3_PalindromeNumber.py
class Solution(object):
    def isPalindrome(self, x):
        """
        :type x: int
        :rtype: bool
        """
        # Negative numbers are not palindromes
        # Numbers ending with 0 (except 0 itself) can't be palindromes
        if x < 0 or (x % 10 == 0 and x != 0):
            return False
        
        reversed_half = 0  # to store the reversed half of the number
        
        # Reverse only half of the digits
        while x > reversed_half:
            digit = x % 10                      # get the last digit
            reversed_half = reversed_half * 10 + digit  # add digit to reversed half
            x //= 10                            # remove the last digit from x
        
        # For even digits: x == reversed_half
        # For odd digits: ignore middle digit by dividing reversed_half by 10
        return x == reversed_half or x == reversed_half // 10
